sdk/trace/batch_span_processor.go

```diff
@@ -167,53 +156,100 @@ func WithBlocking() BatchSpanProcessorOption {
 	}
 }
 
-// processQueue removes spans from the `queue` channel until there is
-// no more data.  It calls the exporter in batches of up to
-// MaxExportBatchSize until all the available data have been processed.
-func (bsp *BatchSpanProcessor) processQueue(batch *[]*export.SpanData) {
+// processQueue removes spans from the `queue` channel until processor
+// is shut down. It calls the exporter in batches of up to MaxExportBatchSize
+// waiting up to ScheduledDelayMillis to form a batch.
+func (bsp *BatchSpanProcessor) processQueue() {
+	timer := time.NewTimer(bsp.o.ScheduledDelayMillis)
+	defer timer.Stop()
+
+	batch := make([]*export.SpanData, 0, bsp.o.MaxExportBatchSize)
+
+	exportSpans := func() {
+		timer.Reset(bsp.o.ScheduledDelayMillis)
+
+		if len(batch) > 0 {
+			bsp.e.ExportSpans(context.Background(), batch)
+			batch = batch[:0]
+		}
+	}
+
+loop:
 	for {
-		// Read spans until either the buffer fills or the
-		// queue is empty.
-		for ok := true; ok && len(*batch) < bsp.o.MaxExportBatchSize; {
-			select {
-			case sd := <-bsp.queue:
-				if sd != nil && sd.SpanContext.IsSampled() {
-					*batch = append(*batch, sd)
+		select {
+		case <-bsp.stopCh:
+			break loop
+		case <-timer.C:
+			exportSpans()
+		case sd := <-bsp.queue:
+			batch = append(batch, sd)
+			if len(batch) == bsp.o.MaxExportBatchSize {
+				if !timer.Stop() {
+					<-timer.C
 				}
-			default:
-				ok = false
+				exportSpans()
 			}
 		}
+	}
 
-		if len(*batch) == 0 {
-			return
+	go func() {
+		bsp.enqueueWait.Wait()
+		close(bsp.queue)
+	}()
+
+	for {
+		if !timer.Stop() {
+			<-timer.C
 		}
 
-		// Send one batch, then continue reading until the
-		// buffer is empty.
-		bsp.e.ExportSpans(context.Background(), *batch)
-		*batch = (*batch)[:0]
+		// This is not needed normally, but use some timeout so we are not stuck
+		// waiting for enqueueWait forever.
+		const waitTimeout = 30 * time.Second
+		timer.Reset(waitTimeout)
+
+		select {
+		case sd := <-bsp.queue:
+			if sd == nil { // queue is closed
+				exportSpans()
+				return
+			}
+
+			batch = append(batch, sd)
+			if len(batch) == bsp.o.MaxExportBatchSize {
+				exportSpans()
+			}
+		case <-timer.C:
+			//TODO: use error callback - see issue #174
+			log.Println("bsp.enqueueWait timeout")
+			exportSpans()
+			return
+		}
 	}
 }
 
 func (bsp *BatchSpanProcessor) enqueue(sd *export.SpanData) {
+	if !sd.SpanContext.IsSampled() {
+		return
+	}
+
+	bsp.enqueueWait.Add(1)
+
 	select {
 	case <-bsp.stopCh:
+		bsp.enqueueWait.Done()
```

Comment:
Use `defer bsp.enqueueWait.Done()` instead of two calls to Done() in this function.